_drafts/heartbeats.md

```diff
@@ -0,0 +1,926 @@
+linearly. We see at 11:30 the device was connected to Wi-Fi, then disconnected
+at 11:45, so the time connected this hour was 15 minutes.
+
+Let's present a problem that happens all the time with embedded systems.
+
+```
+[E][11:30:00] Wi-Fi connected
+
+# Wi-Fi disconnect event lost by device or in transport
+
+[I][12:30:00] Wi-Fi connected
+```
+
+Oh no! A dropped log message.
+
+Can we compute accurately how long the device was connected to Wi-Fi given the
+information above? **No we can't**. We'd probably throw away this event during
+processing or accidentally include it and say Wi-Fi was connected for the entire
+hour.
+
+We can do better.
+
+### Do We Need Both?
+
+Logs and metrics ultimately serve different purposes. You need logs to debug
+hard to reproduce issues and to verify if a particular device has hardware
+issues, and you need metrics to make informed product and business decisions.
+
+**Whether you transform logs into metrics or collect metrics directly in the
+manners mentioned in this article is up to you.**
+
+## Collecting Data from Devices
+
+Embedded systems have quirks about them that make then more difficult to track
+than web applications or mobile phones. Connectivity is slow and unreliable,
+timestamps are rarely accurate if they exist at all, and storage, CPU, RAM,
+power, and bandwidth are all limited.
+
+Due to these constraints, the solutions for how to capture, send, and receive
+data from embedded systems tend to look very different compared to their
+software counterparts.
+
+Optimally, we want to build a metric and monitoring solution that can provide
+all of the following benefits:
+
+- Easy to implement
+- Can measure "health" easily across the entire fleet, different firmware
+  versions, and specific devices
+- Minimal data, bandwidth and connectivity requirements
+- Clock time not required
+- Cost-effective infrastructure
+- Useful even with 5 devices and can scale to 1000's+
+
+We will go through each common approach to gathering monitoring data from
+devices and asses them on the above requirements.
+
+### Plain Text Logging
+
+The first solution that most companies deploy for remote device monitoring is to
+piggy-back off of their existing UART-style logging setup that developers used
+for local debugging. The programmer writes various logs that they think are
+important.
+
+```
+[E][1598845590] Flash sector write error: 6
+[W][1598845590] Wi-Fi disconnected, reason: -2
+[I][1598845592] Battery status: 67%, 3574 mV
+```
+
+The device firmware generates these logs and writes them to a RAM buffer and
+eventually flushed to persistent storage such as NOR flash in a circular buffer
+fashion. Periodically, these logs are vacuumed up by the system and pushed (or
+pulled) to a central location,
+
+Every so often, a single device's logs might be dug up by the engineer who has
+written a `pull-logs-from-s3.py` to track down a bug that a customer had
+reported.
```

Comment:
Since the pull-logs-from-s3 line is already above may change wording slightly
```suggestion
Every so often, a single device's logs might be manually dug up by the engineer to track down a bug that a customer had
reported.
```